# Analyse des resultats des test de différentes structures de UNets sur le dataset Cifar-10

#### Exemple structure UNets :
![unet](../images/myunet.png)

### Informations 

&emsp;Dans ce notebook, nous allons analyser les resultats obtenus sur les tests de différentes structures et hyperparametres de UNets. Nous avons mené des tests pour des UNets de tailles variants entre 3 et 11 couches cahchées de type Conv2D (le modèle à 11 couches étant le plus gros qu'on ai pu tester), en plus des différentes couches qu'on aurait ajouté pour étudier leurs effet (Maxpool, dropout etc..)

## Preparation des outils

In [1]:
import pandas as pd
from IPython.display import display
import numpy as np
from PIL import Image

pd.set_option('display.max_columns', None)

# --------------------------------------------------------------------------------------------------------------

# UNets Sans régularisation

### Tests réalisés :
&emsp;Pour ces premiers tests, nous avons testé des modèles de UNets constitués de la manière suivante:
        - Tailles de 3 à 11 couches cachées
        - Tests avec un nombre d'epochs fixe (40 epochs)
        - Pas de régularisation
        - Différentes fonctions d'activation pour les couches cachées (relu, selu, softmax, softplus)
        - Différentes couches de connexion pour créer la structure UNet (Layers de type Average, Add) 
        
        
&emsp;Nous avons ensuite enregistré les résultats obtenus dans des fichiers CSV, en plus des logs générés par tensorflow, pour étudier les résultats obtenus.
    
&emsp;Nous allons donc charger ces statistiques et voir ce que ce genre de modèles peut donner sur le Dataset v

## Chargement des résultats obtenus sur les modèles avec des couches Average et Avec couche Add:

In [2]:
root_csv_folder = "..\\historique_tests\\"

results_unet_avg = pd.read_csv("{}tested_unet_history.csv".format(root_csv_folder), sep=";")
results_unet_avg = results_unet_avg.sort_values("accuracy_val",ascending=False)

results_unet_add = pd.read_csv("{}tested_unet_add_history.csv".format(root_csv_folder), sep=";")
results_unet_add = results_unet_add.sort_values("accuracy_val",ascending=False)

#### Affichage et premiere analyse des résultats :

In [3]:
display(results_unet_avg)
display(results_unet_add)

nb_couches filter  kernel input_shape activation_couches  \
9           32    3 3      32     32 32 3               selu   
11          32    3 3      32     32 32 3               selu   
11          32    3 3      32     32 32 3               selu   
9           32    3 3      32     32 32 3            softmax   
9           32    3 3      32     32 32 3               selu   
3           32    3 3      32     32 32 3            softmax   
7           32    3 3      32     32 32 3               selu   
5           32    3 3      32     32 32 3            softmax   
9           32    3 3      32     32 32 3            softmax   
9           32    3 3      32     32 32 3            softmax   
9           32    3 3      32     32 32 3            softmax   
5           32    3 3      32     32 32 3               selu   
9           32    3 3      32     32 32 3            softmax   
11          32    3 3      32     32 32 3            softmax   
7           32    3 3      32     32 32 3           softplus   
9           32    3 3      32     32 32 3               relu   
11          32    3 3      32     32 32 3           softplus   
3           32    3 3      32     32 32 3           softplus   
7           32    3 3      32     32 32 3               relu   
5           32    3 3      32     32 32 3               relu   
9           32    3 3      32     32 32 3               relu   
5           32    3 3      32     32 32 3           softplus   
3           32    3 3      32     32 32 3               relu   

   activation_output  maxpool  indexes_maxpool  dropout  indexes_dropout  \
9            softmax    False              NaN    False              NaN   
11           softmax    False              NaN    False              NaN   
11           softmax    False              NaN    False              NaN   
9            softmax    False              NaN    False              NaN   
9            softmax    False              NaN    False              NaN   
3            softmax    False              NaN    False              NaN   
7            softmax    False              NaN    False              NaN   
5            softmax    False              NaN    False              NaN   
9            softmax    False              NaN    False              NaN   
9            softmax    False              NaN    False              NaN   
9            softmax    False              NaN    False              NaN   
5            softmax    False              NaN    False              NaN   
9            softmax    False              NaN    False              NaN   
11           softmax    False              NaN    False              NaN   
7            softmax    False              NaN    False              NaN   
9            softmax    False              NaN    False              NaN   
11           softmax    False              NaN    False              NaN   
3            softmax    False              NaN    False              NaN   
7            softmax    False              NaN    False              NaN   
5            softmax    False              NaN    False              NaN   
9            softmax    False              NaN    False              NaN   
5            softmax    False              NaN    False              NaN   
3            softmax    False              NaN    False              NaN   

    dropout_value  l1l2_couches  l1l2_output  l1_value  l2_value  \
9             0.0         False        False       0.0       0.0   
11            0.0         False        False       0.0       0.0   
11            0.0         False        False       0.0       0.0   
9             0.0         False        False       0.0       0.0   
9             0.0         False        False       0.0       0.0   
3             0.0         False        False       0.0       0.0   
7             0.0         False        False       0.0       0.0   
5             0.0         False        False       0.0       0.0   
9             0.0         Fals

nb_couches filter  kernel input_shape activation_couches activation_output  \
7          32    3 3      32     32 32 3               selu           softmax   
9          32    3 3      32     32 32 3            softmax           softmax   
3          32    3 3      32     32 32 3            softmax           softmax   
3          32    3 3      32     32 32 3               selu           softmax   
7          32    3 3      32     32 32 3           softplus           softmax   
5          32    3 3      32     32 32 3           softplus           softmax   
5          32    3 3      32     32 32 3           softplus           softmax   
5          32    3 3      32     32 32 3               relu           softmax   
7          32    3 3      32     32 32 3               relu           softmax   
3          32    3 3      32     32 32 3           softplus           softmax   

   maxpool  indexes_maxpool  dropout  indexes_dropout  dropout_value  \
7    False              NaN    False              NaN            0.0   
9    False              NaN    False              NaN            0.0   
3    False              NaN    False              NaN            0.0   
3    False              NaN    False              NaN            0.0   
7    False              NaN    False              NaN            0.0   
5    False              NaN    False              NaN            0.0   
5    False              NaN    False              NaN            0.0   
5    False              NaN    False              NaN            0.0   
7    False              NaN    False              NaN            0.0   
3    False              NaN    False              NaN            0.0   

   l1l2_couches  l1l2_output  l1_value  l2_value  indexes_l1l2  \
7         False        False       0.0       0.0           NaN   
9         False        False       0.0       0.0           NaN   
3         False        False       0.0       0.0           NaN   
3         False        False       0.0       0.0           NaN   
7         False        False       0.0       0.0           NaN   
5         False        False       0.0       0.0           NaN   
5         False        False       0.0       0.0           NaN   
5         False        False       0.0       0.0           NaN   
7         False        False       0.0       0.0           NaN   
3         False        False       0.0       0.0           NaN   

                              loss optimizer                      metrics  \
7  sparse_categorical_crossentropy      Adam  sparse_categorical_accuracy   
9  sparse_categorical_crossentropy      Adam  sparse_categorical_accuracy   
3  sparse_categorical_crossentropy      Adam  sparse_categorical_accuracy   
3  sparse_categorical_crossentropy      Adam  sparse_categorical_accuracy   
7  sparse_categorical_crossentropy      Adam  sparse_categorical_accuracy   
5  sparse_categorical_crossentropy      Adam  sparse_categorical_accuracy   
5  sparse_categorical_crossentropy      Adam  sparse_categorical_accuracy   
5  sparse_categorical_crossentropy      Adam  sparse_categorical_accuracy   
7  sparse_categorical_crossentropy      Adam  sparse_categorical_accuracy   
3  sparse_categorical_crossentropy      Adam  sparse_categorical_accuracy   

  padding  epochs  model_id  accuracy_train  accuracy_val   dossier  
7    same      40   3568266         0.95096        0.5813  20200126  
9    same      40   9230533         0.99646        0.5705  20200126  
3    same      40   3696804         0.95534        0.5675  20200126  
3    same      40   7915611         0.89546        0.5388  20200126  
7    same      40   1500229         0.97280        0.5059  20200126  
5    same      40   3978437         0.96484        0.4805  20200126  
5    same      40    160055         0.96818        0.4721  20200126  
5    same      40   3005369         0.95494        0.4609  20200126  
7    same      40   6522794         0.94068        0.4586  20200126  
3    same      40    485416         0.95150

##### Premières Constatations:
&emsp;La première constatation qu'on peut faire, c'est que les résultats obtenus avec les fonctions d'activation "relu" et "softplus", que ce soit sur les modèles avec des couches Average ou des couches Add, ne sont pas très intéressantes comparées au fonctions "selu" et "softplus": pour les deux première des résultats qui tournent autours des 50% sur la validation, contre 58% environs sur les deux dernières.
    
&emsp;Nous allons donc nous focaliser sur les fonctions d'activation "softplus" et "selu" dans nos prochains tests et analyses.

In [4]:
# Filtre sur les fonctions d'activation et affichage des resultats
results_unet_avg = results_unet_avg[results_unet_avg.activation_couches.isin(["selu", "softplus"])]
results_unet_add = results_unet_add[results_unet_add.activation_couches.isin(["selu", "softplus"])]

display(results_unet_add)
display(results_unet_avg)

nb_couches filter  kernel input_shape activation_couches activation_output  \
7          32    3 3      32     32 32 3               selu           softmax   
3          32    3 3      32     32 32 3               selu           softmax   
7          32    3 3      32     32 32 3           softplus           softmax   
5          32    3 3      32     32 32 3           softplus           softmax   
5          32    3 3      32     32 32 3           softplus           softmax   
3          32    3 3      32     32 32 3           softplus           softmax   

   maxpool  indexes_maxpool  dropout  indexes_dropout  dropout_value  \
7    False              NaN    False              NaN            0.0   
3    False              NaN    False              NaN            0.0   
7    False              NaN    False              NaN            0.0   
5    False              NaN    False              NaN            0.0   
5    False              NaN    False              NaN            0.0   
3    False              NaN    False              NaN            0.0   

   l1l2_couches  l1l2_output  l1_value  l2_value  indexes_l1l2  \
7         False        False       0.0       0.0           NaN   
3         False        False       0.0       0.0           NaN   
7         False        False       0.0       0.0           NaN   
5         False        False       0.0       0.0           NaN   
5         False        False       0.0       0.0           NaN   
3         False        False       0.0       0.0           NaN   

                              loss optimizer                      metrics  \
7  sparse_categorical_crossentropy      Adam  sparse_categorical_accuracy   
3  sparse_categorical_crossentropy      Adam  sparse_categorical_accuracy   
7  sparse_categorical_crossentropy      Adam  sparse_categorical_accuracy   
5  sparse_categorical_crossentropy      Adam  sparse_categorical_accuracy   
5  sparse_categorical_crossentropy      Adam  sparse_categorical_accuracy   
3  sparse_categorical_crossentropy      Adam  sparse_categorical_accuracy   

  padding  epochs  model_id  accuracy_train  accuracy_val   dossier  
7    same      40   3568266         0.95096        0.5813  20200126  
3    same      40   7915611         0.89546        0.5388  20200126  
7    same      40   1500229         0.97280        0.5059  20200126  
5    same      40   3978437         0.96484        0.4805  20200126  
5    same      40    160055         0.96818        0.4721  20200126  
3    same      40    485416         0.95150        0.4330  20200126

nb_couches filter  kernel input_shape activation_couches  \
9           32    3 3      32     32 32 3               selu   
11          32    3 3      32     32 32 3               selu   
11          32    3 3      32     32 32 3               selu   
9           32    3 3      32     32 32 3               selu   
7           32    3 3      32     32 32 3               selu   
5           32    3 3      32     32 32 3               selu   
7           32    3 3      32     32 32 3           softplus   
11          32    3 3      32     32 32 3           softplus   
3           32    3 3      32     32 32 3           softplus   
5           32    3 3      32     32 32 3           softplus   

   activation_output  maxpool  indexes_maxpool  dropout  indexes_dropout  \
9            softmax    False              NaN    False              NaN   
11           softmax    False              NaN    False              NaN   
11           softmax    False              NaN    False              NaN   
9            softmax    False              NaN    False              NaN   
7            softmax    False              NaN    False              NaN   
5            softmax    False              NaN    False              NaN   
7            softmax    False              NaN    False              NaN   
11           softmax    False              NaN    False              NaN   
3            softmax    False              NaN    False              NaN   
5            softmax    False              NaN    False              NaN   

    dropout_value  l1l2_couches  l1l2_output  l1_value  l2_value  \
9             0.0         False        False       0.0       0.0   
11            0.0         False        False       0.0       0.0   
11            0.0         False        False       0.0       0.0   
9             0.0         False        False       0.0       0.0   
7             0.0         False        False       0.0       0.0   
5             0.0         False        False       0.0       0.0   
7             0.0         False        False       0.0       0.0   
11            0.0         False        False       0.0       0.0   
3             0.0         False        False       0.0       0.0   
5             0.0         False        False       0.0       0.0   

    indexes_l1l2                             loss optimizer  \
9            NaN  sparse_categorical_crossentropy      Adam   
11           NaN  sparse_categorical_crossentropy      Adam   
11           NaN  sparse_categorical_crossentropy      Adam   
9            NaN  sparse_categorical_crossentropy      Adam   
7            NaN  sparse_categorical_crossentropy      Adam   
5            NaN  sparse_categorical_crossentropy      Adam   
7            NaN  sparse_categorical_crossentropy      Adam   
11           NaN  sparse_categorical_crossentropy      Adam   
3            NaN  sparse_categorical_crossentropy      Adam   
5            NaN  sparse_categorical_crossentropy      Adam   

                        metrics padding  epochs  model_id  accuracy_train  \
9   sparse_categorical_accuracy    same      40   7318436         0.95690   
11  sparse_categorical_accuracy    same      40   3928073         0.95738   
11  sparse_categorical_accuracy    same      40   6067132         0.95094   
9   sparse_categorical_accuracy    same      40   3589055         0.94418   
7   sparse_categorical_accuracy    same      40   6561474         0.92546   
5   sparse_categorical_accuracy    same      40   7982485         0.95870   
7   sparse_categorical_accuracy    same      40   3979387         0.98030   
11  sparse_categorical_accuracy    same      40   2367066         0.97968   
3   sparse_categorical_accuracy    same      40   1360206         0.99472   
5   sparse_categorical_accuracy    same      40   9317203         0.95512   

    accuracy_val   dossier  
9         0.6059  20200126  
11        0.5968  20200127  
11        0.5963  20200127  
9         0.5925  20200127  
7         0.5890  20200126

##### Deuxièmes Constatations:
&emsp;Ici, on peut faire deux constations de plus:

&emsp;- En comparant les résultats obtenus pour les modèles utilisant des couches Average et des modèles avec des couches Add, on se rend compte que le premier type de modèles est bien plus interessants que le deuxième en terme de résultats (avec des 
add, on tourne autours des 50% sur la validation, contre près de 60% avec des Average.

&emsp;- On remarque aussi que les meilleurs modèles sont ceux qui utilisent la fonction d'activation "selu".
    
&emsp;Nous allons donc continuer les tests sur des modèles utilisant des couches Average et la fonction d'activation "selu".

In [5]:
results_unet_avg = results_unet_avg[results_unet_avg.activation_couches == "selu"]

display(results_unet_avg)

nb_couches filter  kernel input_shape activation_couches  \
9           32    3 3      32     32 32 3               selu   
11          32    3 3      32     32 32 3               selu   
11          32    3 3      32     32 32 3               selu   
9           32    3 3      32     32 32 3               selu   
7           32    3 3      32     32 32 3               selu   
5           32    3 3      32     32 32 3               selu   

   activation_output  maxpool  indexes_maxpool  dropout  indexes_dropout  \
9            softmax    False              NaN    False              NaN   
11           softmax    False              NaN    False              NaN   
11           softmax    False              NaN    False              NaN   
9            softmax    False              NaN    False              NaN   
7            softmax    False              NaN    False              NaN   
5            softmax    False              NaN    False              NaN   

    dropout_value  l1l2_couches  l1l2_output  l1_value  l2_value  \
9             0.0         False        False       0.0       0.0   
11            0.0         False        False       0.0       0.0   
11            0.0         False        False       0.0       0.0   
9             0.0         False        False       0.0       0.0   
7             0.0         False        False       0.0       0.0   
5             0.0         False        False       0.0       0.0   

    indexes_l1l2                             loss optimizer  \
9            NaN  sparse_categorical_crossentropy      Adam   
11           NaN  sparse_categorical_crossentropy      Adam   
11           NaN  sparse_categorical_crossentropy      Adam   
9            NaN  sparse_categorical_crossentropy      Adam   
7            NaN  sparse_categorical_crossentropy      Adam   
5            NaN  sparse_categorical_crossentropy      Adam   

                        metrics padding  epochs  model_id  accuracy_train  \
9   sparse_categorical_accuracy    same      40   7318436         0.95690   
11  sparse_categorical_accuracy    same      40   3928073         0.95738   
11  sparse_categorical_accuracy    same      40   6067132         0.95094   
9   sparse_categorical_accuracy    same      40   3589055         0.94418   
7   sparse_categorical_accuracy    same      40   6561474         0.92546   
5   sparse_categorical_accuracy    same      40   7982485         0.95870   

    accuracy_val   dossier  
9         0.6059  20200126  
11        0.5968  20200127  
11        0.5963  20200127  
9         0.5925  20200127  
7         0.5890  20200126  
5         0.5772  20200125

###  Conclusion premiers tests:
&emsp;Ces premiers tests nous ont permis de cibler le type de modèle le plus interessant: Modèle utilisant des couches Average et une fonction d'activation "selu" pour ses couches cachées dont les résultats tournent autours des 60% en validation sans tenir compte de la taille du modèle.


&emsp;On remarque aussi que tous ces modèles overfittent très fortement, en arrivant à 95% en train. Nous allons donc poursuivre les tests sur ce type de modèles en utilisant des techniques de régularisation.

# --------------------------------------------------------------------------------------------------------------

# UNets Avec régularisation :

### Tests réalisés :
&emsp;Pour nouveaux tests, nous avons testé des modèles de UNets constitués de la manière suivante:
        - Tailles de 3 à 11 couches cachées
        - Tests avec un nombre d'epochs fixe (40 epochs)
        - Utilisation de techniques de régularisation 
        - Fonctions d'activation pour les couches cachées "selu"
        - Couches de connexion de type Average pour créer la structure UNet 
        
        
&emsp;Comme pour les premiers tests, nous avons ensuite enregistré les résultats obtenus dans des fichiers CSV, en plus des logs générés par tensorflow, et nous allons charger ces statistiques et voir ce que ce genre de modèles peut donner.

#### I) Modèles de différentes tailles avec régularisation L2:

##### Partie 1:

In [6]:
root_csv_folder = "..\\historique_tests\\"

# Organisation sur la valeur de la L2
results_unet_avg_selu_l2 = pd.read_csv("{}tested_unet_selu_l2_history.csv".format(root_csv_folder), sep=";")
results_unet_avg_selu_l2 = results_unet_avg_selu_l2.sort_values("l2_value",ascending=True)

display(results_unet_avg_selu_l2)

nb_couches filter  kernel input_shape activation_couches activation_output  \
3          32    3 3      32     32 32 3               selu           softmax   
3          32    3 3      32     32 32 3               selu           softmax   
3          32    3 3      32     32 32 3               selu           softmax   
3          32    3 3      32     32 32 3               selu           softmax   
3          32    3 3      32     32 32 3               selu           softmax   
3          32    3 3      32     32 32 3               selu           softmax   
3          32    3 3      32     32 32 3               selu           softmax   
3          32    3 3      32     32 32 3               selu           softmax   
9          32    3 3      32     32 32 3               selu           softmax   
9          32    3 3      32     32 32 3               selu           softmax   
9          32    3 3      32     32 32 3               selu           softmax   
3          32    3 3      32     32 32 3               selu           softmax   
3          32    3 3      32     32 32 3               selu           softmax   
9          32    3 3      32     32 32 3               selu           softmax   
7          32    3 3      32     32 32 3               selu           softmax   
7          32    3 3      32     32 32 3               selu           softmax   
7          32    3 3      32     32 32 3               selu           softmax   
3          32    3 3      32     32 32 3               selu           softmax   
3          32    3 3      32     32 32 3               selu           softmax   
3          32    3 3      32     32 32 3               selu           softmax   
3          32    3 3      32     32 32 3               selu           softmax   

   maxpool  indexes_maxpool  dropout  indexes_dropout  dropout_value  \
3    False              NaN    False              NaN            0.0   
3    False              NaN    False              NaN            0.0   
3    False              NaN    False              NaN            0.0   
3    False              NaN    False              NaN            0.0   
3    False              NaN    False              NaN            0.0   
3    False              NaN    False              NaN            0.0   
3    False              NaN    False              NaN            0.0   
3    False              NaN    False              NaN            0.0   
9    False              NaN    False              NaN            0.0   
9    False              NaN    False              NaN            0.0   
9    False              NaN    False              NaN            0.0   
3    False              NaN    False              NaN            0.0   
3    False              NaN    False              NaN            0.0   
9    False              NaN    False              NaN            0.0   
7    False              NaN    False              NaN            0.0   
7    False              NaN    False              NaN            0.0   
7    False              NaN    False              NaN            0.0   
3    False              NaN    False              NaN            0.0   
3    False              NaN    False              NaN            0.0   
3    False              NaN    False              NaN            0.0   
3    False              NaN    False              NaN            0.0   

   l1l2_couches  l1l2_output  l1_value  l2_value       indexes_l1l2  \
3          True        False       0.0     0.001              1 2 3   
3          True        False       0.0     0.001              1 2 3   
3          True        False       0.0     0.001              1 2 3   
3          True        False       0.0     0.010              1 2 3   
3          True        False       0.0     0.010              1 2 3   
3          True        False       0.0     0.010              1 2 3   
3          True         True       0.0     0.010              1 2 3   
3          True         True       0.0     0.010              1 2 3

#### Constatations:
&emsp;Pour ces tests, nous avons testé différentes valeurs pour la régularisation l2: 0.001, 0.01, 0.1, 0.02, 0.5 et 1.

&emsp;Au début, nous avons lancé des tests avec des valeurs pour la L2 assez petites, mais nous avons été surpris de remarquer que ces petites valeurs n'avaient pas beaucoup d'effet de régularisation sur notre modèle. 

&emsp;Nous avons lancé des tests avec des valeurs plus grands (0.5 ou encore 1), et effectivement, ces grandes valeurs régularisent beaucoup mieux. Par exemple sur un seul type de modèle (même nombre de couches), avec une valeur de 0.001, on a toujours plus de 90% de résultat sur le train, alors qu'avec une valeur de 0.5, on arrive à descendre à 72% sur le train, avec toujours 59% sur la validation. 

&emsp;Nous allons donc jouer sur ces valeurs pour la suite, mais sur de plus gros modèles (étant donné que la majorité de cette vague de tests a été réalisée sur des modèles à 3 couches cachées). 

##### Partie 2:

&emsp;Une fois ces propriétés découverts, nous avons décidé de tester avec cette valeur de L2 des modèles plus gros (5, 7 et 9 couches).

In [7]:
root_csv_folder = "..\\historique_tests\\"

# Organisation sur la valeur de la L2
results_unet_l2 = pd.read_csv("{}tested_unet_l2_history.csv".format(root_csv_folder), sep=";")
results_unet_l2 = results_unet_l2.sort_values("l2_value",ascending=True)

display(results_unet_l2)

nb_couches filter  kernel input_shape activation_couches activation_output  \
5          32    3 3      32     32 32 3               selu           softmax   
5          32    3 3      32     32 32 3               selu           softmax   
5          32    3 3      32     32 32 3               selu           softmax   
7          32    3 3      32     32 32 3               selu           softmax   
7          32    3 3      32     32 32 3               selu           softmax   
7          32    3 3      32     32 32 3               selu           softmax   
9          32    3 3      32     32 32 3               selu           softmax   
9          32    3 3      32     32 32 3               selu           softmax   
9          32    3 3      32     32 32 3               selu           softmax   

   maxpool  indexes_maxpool  dropout  indexes_dropout  dropout_value  \
5    False              NaN    False              NaN            0.0   
5    False              NaN    False              NaN            0.0   
5    False              NaN    False              NaN            0.0   
7    False              NaN    False              NaN            0.0   
7    False              NaN    False              NaN            0.0   
7    False              NaN    False              NaN            0.0   
9    False              NaN    False              NaN            0.0   
9    False              NaN    False              NaN            0.0   
9    False              NaN    False              NaN            0.0   

   l1l2_couches  l1l2_output  l1_value  l2_value     indexes_l1l2  \
5          True        False       0.0       0.5        1 2 3 4 5   
5          True        False       0.0       0.5        1 2 3 4 5   
5          True        False       0.0       0.5        1 2 3 4 5   
7          True        False       0.0       0.5    1 2 3 4 5 6 7   
7          True        False       0.0       0.5    1 2 3 4 5 6 7   
7          True        False       0.0       0.5    1 2 3 4 5 6 7   
9          True        False       0.0       0.5  1 2 3 4 5 6 7 9   
9          True        False       0.0       0.5  1 2 3 4 5 6 7 9   
9          True        False       0.0       0.5  1 2 3 4 5 6 7 9   

                              loss optimizer                      metrics  \
5  sparse_categorical_crossentropy      Adam  sparse_categorical_accuracy   
5  sparse_categorical_crossentropy      Adam  sparse_categorical_accuracy   
5  sparse_categorical_crossentropy      Adam  sparse_categorical_accuracy   
7  sparse_categorical_crossentropy      Adam  sparse_categorical_accuracy   
7  sparse_categorical_crossentropy      Adam  sparse_categorical_accuracy   
7  sparse_categorical_crossentropy      Adam  sparse_categorical_accuracy   
9  sparse_categorical_crossentropy      Adam  sparse_categorical_accuracy   
9  sparse_categorical_crossentropy      Adam  sparse_categorical_accuracy   
9  sparse_categorical_crossentropy      Adam  sparse_categorical_accuracy   

  padding  epochs  model_id  accuracy_train  accuracy_val   dossier  
5    same      40    143471         0.66574        0.5524  20200202  
5    same      40   6338920         0.70396        0.5608  20200202  
5    same      40   9106991         0.65898        0.5525  20200202  
7    same      40   3966471         0.64230        0.5408  20200202  
7    same      40   2780420         0.72756        0.5779  20200202  
7    same      40    926451         0.71426        0.5959  20200202  
9    same      40   9438031         0.69784        0.5595  20200202  
9    same      40   6372702         0.69508        0.5457  20200202  
9    same      40   3822422         0.70496        0.5488  20200202

In [8]:
display(results_unet_l2[["nb_couches", "l1l2_couches", "indexes_l1l2","l2_value", "accuracy_train", "accuracy_val"]])

nb_couches  l1l2_couches     indexes_l1l2  l2_value  accuracy_train  \
5          32          True        1 2 3 4 5       0.5         0.66574   
5          32          True        1 2 3 4 5       0.5         0.70396   
5          32          True        1 2 3 4 5       0.5         0.65898   
7          32          True    1 2 3 4 5 6 7       0.5         0.64230   
7          32          True    1 2 3 4 5 6 7       0.5         0.72756   
7          32          True    1 2 3 4 5 6 7       0.5         0.71426   
9          32          True  1 2 3 4 5 6 7 9       0.5         0.69784   
9          32          True  1 2 3 4 5 6 7 9       0.5         0.69508   
9          32          True  1 2 3 4 5 6 7 9       0.5         0.70496   

   accuracy_val  
5        0.5524  
5        0.5608  
5        0.5525  
7        0.5408  
7        0.5779  
7        0.5959  
9        0.5595  
9        0.5457  
9        0.5488

#### Constatations:
&emsp;Le résultats obtenus avec ce genre de modèles ne sont pas vraiment intéressants, étant donné qu'on ne dépasse pas les 60% (même résultat qu'en validations). On n'a donc aucune plus-value mise à part la réduction de l'overfitiing (de 95% à 70%). 

Exemple courbe d'entrainement sur le modèle à 7 couches :
![courbe unet l2](overfit_unet_l2.png)

#### II) Modèles de différentes tailles avec régularisation Dropout:

In [9]:
root_csv_folder = "..\\historique_tests\\"

# Organisation sur la valeur de la L2
results_unet_drp = pd.read_csv("{}tested_unet_dropout_history.csv".format(root_csv_folder), sep=";")
results_unet_drp = results_unet_drp.sort_values("l2_value",ascending=True)

display(results_unet_drp)

nb_couches filter  kernel input_shape activation_couches activation_output  \
3          32    3 3      32     32 32 3               selu           softmax   
3          32    3 3      32     32 32 3               selu           softmax   
3          32    3 3      32     32 32 3               selu           softmax   
3          32    3 3      32     32 32 3               selu           softmax   
3          32    3 3      32     32 32 3               selu           softmax   
3          32    3 3      32     32 32 3               selu           softmax   
3          32    3 3      32     32 32 3               selu           softmax   
3          32    3 3      32     32 32 3               selu           softmax   

   maxpool  indexes_maxpool  dropout indexes_dropout  dropout_value  \
3    False              NaN     True               0            0.2   
3    False              NaN     True               0            0.2   
3    False              NaN     True               0            0.4   
3    False              NaN     True               0            0.4   
3    False              NaN     True               1            0.2   
3    False              NaN     True               1            0.2   
3    False              NaN     True             0 1            0.2   
3    False              NaN     True             0 1            0.2   

   l1l2_couches  l1l2_output  l1_value  l2_value  indexes_l1l2  \
3         False        False       0.0       0.0           NaN   
3         False        False       0.0       0.0           NaN   
3         False        False       0.0       0.0           NaN   
3         False        False       0.0       0.0           NaN   
3         False        False       0.0       0.0           NaN   
3         False        False       0.0       0.0           NaN   
3         False        False       0.0       0.0           NaN   
3         False        False       0.0       0.0           NaN   

                              loss optimizer                      metrics  \
3  sparse_categorical_crossentropy      Adam  sparse_categorical_accuracy   
3  sparse_categorical_crossentropy      Adam  sparse_categorical_accuracy   
3  sparse_categorical_crossentropy      Adam  sparse_categorical_accuracy   
3  sparse_categorical_crossentropy      Adam  sparse_categorical_accuracy   
3  sparse_categorical_crossentropy      Adam  sparse_categorical_accuracy   
3  sparse_categorical_crossentropy      Adam  sparse_categorical_accuracy   
3  sparse_categorical_crossentropy      Adam  sparse_categorical_accuracy   
3  sparse_categorical_crossentropy      Adam  sparse_categorical_accuracy   

  padding  epochs  model_id  accuracy_train  accuracy_val   dossier  
3    same      40   5936336         0.50860        0.4081  20200127  
3    same      40    973386         0.56624        0.4508  20200127  
3    same      40   8447412         0.22334        0.2089  20200127  
3    same      40   5326246         0.21740        0.1960  20200127  
3    same      40   1829193         0.95690        0.5819  20200127  
3    same      40   5448510         0.96206        0.5863  20200127  
3    same      40   5239371         0.53534        0.4425  20200128  
3    same      40   6837419         0.48536        0.4108  20200128

In [10]:
display(results_unet_drp[["nb_couches", "dropout", "indexes_dropout", "dropout_value", "accuracy_train", "accuracy_val"]])

nb_couches  dropout indexes_dropout  dropout_value  accuracy_train  \
3          32     True               0            0.2         0.50860   
3          32     True               0            0.2         0.56624   
3          32     True               0            0.4         0.22334   
3          32     True               0            0.4         0.21740   
3          32     True               1            0.2         0.95690   
3          32     True               1            0.2         0.96206   
3          32     True             0 1            0.2         0.53534   
3          32     True             0 1            0.2         0.48536   

   accuracy_val  
3        0.4081  
3        0.4508  
3        0.2089  
3        0.1960  
3        0.5819  
3        0.5863  
3        0.4425  
3        0.4108

#### Constatations:
&emsp;D'après ces résultats, On peut considérer que cette technique de dropout ne nous est pas d'une grande utilité. Dans les cas où elle ne pénalise pas trop nos modèles, elle ne nous permet pas d'augmenter notre accuracy en validation ni de diminuer l'overfitting (exemple sur la ligne pour la régularisation sur la couche 1).

#### III) Modèles de différentes tailles avec maxpool:

In [11]:
root_csv_folder = "..\\historique_tests\\"

# Organisation sur la valeur de la L2
results_unet_maxpool = pd.read_csv("{}tested_unet_selu_maxpool_history.csv".format(root_csv_folder), sep=";")
results_unet_maxpool = results_unet_maxpool.sort_values("l2_value",ascending=True)

display(results_unet_maxpool)

nb_couches filter  kernel input_shape activation_couches activation_output  \
3          32    3 3      32     32 32 3               selu           softmax   
3          32    3 3      32     32 32 3               selu           softmax   
3          32    3 3      32     32 32 3               selu           softmax   
5          32    3 3      32     32 32 3               selu           softmax   
5          32    3 3      32     32 32 3               selu           softmax   
5          32    3 3      32     32 32 3               selu           softmax   
7          32    3 3      32     32 32 3               selu           softmax   
7          32    3 3      32     32 32 3               selu           softmax   
7          32    3 3      32     32 32 3               selu           softmax   
9          32    3 3      32     32 32 3               selu           softmax   
9          32    3 3      32     32 32 3               selu           softmax   
9          32    3 3      32     32 32 3               selu           softmax   

   maxpool indexes_maxpool  dropout  indexes_dropout  dropout_value  \
3     True               1    False              NaN            0.0   
3     True               1    False              NaN            0.0   
3     True               1    False              NaN            0.0   
5     True             1 2    False              NaN            0.0   
5     True             1 2    False              NaN            0.0   
5     True             1 2    False              NaN            0.0   
7     True           1 2 3    False              NaN            0.0   
7     True           1 2 3    False              NaN            0.0   
7     True           1 2 3    False              NaN            0.0   
9     True         1 2 3 4    False              NaN            0.0   
9     True         1 2 3 4    False              NaN            0.0   
9     True         1 2 3 4    False              NaN            0.0   

   l1l2_couches  l1l2_output  l1_value  l2_value  indexes_l1l2  \
3         False        False       0.0       0.0           NaN   
3         False        False       0.0       0.0           NaN   
3         False        False       0.0       0.0           NaN   
5         False        False       0.0       0.0           NaN   
5         False        False       0.0       0.0           NaN   
5         False        False       0.0       0.0           NaN   
7         False        False       0.0       0.0           NaN   
7         False        False       0.0       0.0           NaN   
7         False        False       0.0       0.0           NaN   
9         False        False       0.0       0.0           NaN   
9         False        False       0.0       0.0           NaN   
9         False        False       0.0       0.0           NaN   

                              loss optimizer                      metrics  \
3  sparse_categorical_crossentropy      Adam  sparse_categorical_accuracy   
3  sparse_categorical_crossentropy      Adam  sparse_categorical_accuracy   
3  sparse_categorical_crossentropy      Adam  sparse_categorical_accuracy   
5  sparse_categorical_crossentropy      Adam  sparse_categorical_accuracy   
5  sparse_categorical_crossentropy      Adam  sparse_categorical_accuracy   
5  sparse_categorical_crossentropy      Adam  sparse_categorical_accuracy   
7  sparse_categorical_crossentropy      Adam  sparse_categorical_accuracy   
7  sparse_categorical_crossentropy      Adam  sparse_categorical_accuracy   
7  sparse_categorical_crossentropy      Adam  sparse_categorical_accuracy   
9  sparse_categorical_crossentropy      Adam  sparse_categorical_accuracy   
9  sparse_categorical_crossentropy      Adam  sparse_categorical_accuracy   
9  sparse_categorical_crossentropy      Adam  sparse_categorical_accuracy   

  padding  epochs  model_id  accuracy_train  accuracy_val   dossier  
3    same      40   6202514         0.97274        0.5693  20200127  
3    same      40   3398

In [12]:
display(results_unet_maxpool[["nb_couches", "maxpool", "indexes_maxpool", "accuracy_train", "accuracy_val"]])

nb_couches  maxpool indexes_maxpool  accuracy_train  accuracy_val
3          32     True               1         0.97274        0.5693
3          32     True               1         0.78778        0.5297
3          32     True               1         0.91888        0.5549
5          32     True             1 2         0.90578        0.5880
5          32     True             1 2         0.93848        0.5798
5          32     True             1 2         0.93716        0.5805
7          32     True           1 2 3         0.94658        0.6018
7          32     True           1 2 3         0.93796        0.5823
7          32     True           1 2 3         0.92722        0.5810
9          32     True         1 2 3 4         0.94236        0.5896
9          32     True         1 2 3 4         0.92278        0.5871
9          32     True         1 2 3 4         0.94790        0.6027

#### Constatations:
&emsp;En voyant ces résultats, on peut se dire que les maxpool pourraient peut-être présenter un intérêt, vu qu'elle améliore un tout petit peu l'accuracy en validation, mais surtout, étant donné que les modèles overfittent toujours énormément, il peu aussi y avoir un bon potentiel de régularisation et qui pourrait nous donner de meilleurs résultats. Nous allons donc essayer de combiner cette technique à une autre régularisation plus tard.

#### IV) Modèles de différentes tailles avec Dropout et L2:

In [13]:
root_csv_folder = "..\\historique_tests\\"

# Organisation sur la valeur de la L2
results_unet_drp_l2 = pd.read_csv("{}tested_unet_dropout_l2_history.csv".format(root_csv_folder), sep=";")
results_unet_drp_l2 = results_unet_drp_l2.sort_values("l2_value",ascending=True)

display(results_unet_drp_l2)

nb_couches filter  kernel input_shape activation_couches activation_output  \
3          32    3 3      32     32 32 3               selu           softmax   
3          32    3 3      32     32 32 3               selu           softmax   
3          32    3 3      32     32 32 3               selu           softmax   
3          32    3 3      32     32 32 3               selu           softmax   
3          32    3 3      32     32 32 3               selu           softmax   
3          32    3 3      32     32 32 3               selu           softmax   

   maxpool  indexes_maxpool  dropout indexes_dropout  dropout_value  \
3    False              NaN     True               0            0.2   
3    False              NaN     True               0            0.2   
3    False              NaN     True         0 1 2 3            0.2   
3    False              NaN     True         0 1 2 3            0.2   
3    False              NaN     True         0 1 2 3            0.2   
3    False              NaN     True         0 1 2 3            0.2   

   l1l2_couches  l1l2_output  l1_value  l2_value indexes_l1l2  \
3          True        False       0.0       0.5        1 2 3   
3          True        False       0.0       0.5        1 2 3   
3          True        False       0.0       1.0        1 2 3   
3          True        False       0.0       1.0        1 2 3   
3          True        False       0.0       1.0        1 2 3   
3          True        False       0.0       1.0        1 2 3   

                              loss optimizer                      metrics  \
3  sparse_categorical_crossentropy      Adam  sparse_categorical_accuracy   
3  sparse_categorical_crossentropy      Adam  sparse_categorical_accuracy   
3  sparse_categorical_crossentropy      Adam  sparse_categorical_accuracy   
3  sparse_categorical_crossentropy      Adam  sparse_categorical_accuracy   
3  sparse_categorical_crossentropy      Adam  sparse_categorical_accuracy   
3  sparse_categorical_crossentropy      Adam  sparse_categorical_accuracy   

  padding  epochs  model_id  accuracy_train  accuracy_val   dossier  
3    same      40   8436092         0.52374        0.4857  20200128  
3    same      40   2137471         0.54976        0.5035  20200128  
3    same      40    590179         0.54746        0.5219  20200128  
3    same      40   1808046         0.51630        0.4894  20200128  
3    same      40   8636161         0.53728        0.5123  20200128  
3    same      40   4135911         0.53184        0.4943  20200128

In [14]:
display(results_unet_drp_l2[["nb_couches", "dropout", "indexes_dropout", "dropout_value", "l1l2_couches", "indexes_l1l2","l2_value", "accuracy_train", "accuracy_val"]])

nb_couches  dropout indexes_dropout  dropout_value  l1l2_couches  \
3          32     True               0            0.2          True   
3          32     True               0            0.2          True   
3          32     True         0 1 2 3            0.2          True   
3          32     True         0 1 2 3            0.2          True   
3          32     True         0 1 2 3            0.2          True   
3          32     True         0 1 2 3            0.2          True   

  indexes_l1l2  l2_value  accuracy_train  accuracy_val  
3        1 2 3       0.5         0.52374        0.4857  
3        1 2 3       0.5         0.54976        0.5035  
3        1 2 3       1.0         0.54746        0.5219  
3        1 2 3       1.0         0.51630        0.4894  
3        1 2 3       1.0         0.53728        0.5123  
3        1 2 3       1.0         0.53184        0.4943

#### Constatations:
&emsp;La combinaison de ces deux techniques n'est pas très intéressante car on ne dépasse pas les 50%, mais aussi car nos modèles sont trop pénalisés. Nous faisons maintenant face à de l'underfitting comme le montrent ces logs :

![courbe unet drp_l2](courbe_unet_drp_l2.png)

#### V) Modèles de différentes tailles avec Maxpool et L2 / Maxpool avec noyeau de (5,5) au lieu de (3,3):

In [15]:
root_csv_folder = "..\\historique_tests\\"

# Organisation sur la valeur de la L2
results_unet_maxpool_l2 = pd.read_csv("{}tested_unet_maxpool_l2_history.csv".format(root_csv_folder), sep=";")
results_unet_maxpool_l2 = results_unet_maxpool_l2.sort_values("l2_value",ascending=True)

display(results_unet_maxpool_l2)

nb_couches filter  kernel input_shape activation_couches activation_output  \
5          32    5 5      32     32 32 3               selu           softmax   
5          32    3 3      32     32 32 3               selu           softmax   
3          32    3 3      32     32 32 3               selu           softmax   

   maxpool indexes_maxpool  dropout  indexes_dropout  dropout_value  \
5     True             1 2    False              NaN            0.0   
5     True             1 2    False              NaN            0.0   
3     True               1    False              NaN            0.0   

   l1l2_couches  l1l2_output  l1_value  l2_value indexes_l1l2  \
5         False        False       0.0       0.0          NaN   
5          True        False       0.0       0.2    1 2 3 4 5   
3          True        False       0.0       0.2        1 2 3   

                              loss optimizer                      metrics  \
5  sparse_categorical_crossentropy      Adam  sparse_categorical_accuracy   
5  sparse_categorical_crossentropy      Adam  sparse_categorical_accuracy   
3  sparse_categorical_crossentropy      Adam  sparse_categorical_accuracy   

  padding  epochs  model_id  accuracy_train  accuracy_val   dossier  
5    same      40   7784129         0.92016        0.5739  20200209  
5    same      40   1443719         0.73998        0.5265  20200209  
3    same      40   8070740         0.80018        0.5517  20200209

In [16]:
display(results_unet_maxpool_l2[["filter", "maxpool", "indexes_maxpool", "l1l2_couches", "indexes_l1l2","l2_value", "accuracy_train", "accuracy_val"]])

filter  maxpool indexes_maxpool  l1l2_couches indexes_l1l2  l2_value  \
5    5 5     True             1 2         False          NaN       0.0   
5    3 3     True             1 2          True    1 2 3 4 5       0.2   
3    3 3     True               1          True        1 2 3       0.2   

   accuracy_train  accuracy_val  
5         0.92016        0.5739  
5         0.73998        0.5265  
3         0.80018        0.5517

#### Constatations:
&emsp; - Maxpool, L2 : La combinaison de ces deux techniques n'est pas très intéressante car on overfitte toujours et on ne dépasse pas les 55%. De plus, on peut souligner que le modèle à 3 couches est plus performant que le modèle à 5 couches (sur train et validation). nous pensons que cela est du à la taille du kernel.

&emsp; - Maxpool, et kernell de 5*5 : On remarque ici qu'avec un plus gros noyau, nous avons de meilleurs résultats. Nous allons utiliser cette propriété pour créer notre modèle ! 

# --------------------------------------------------------------------------------------------------------------

# Conclusion UNets :

###### &emsp; En essayant avec un kernel de 9*9, on retombe à une accuracy de 48% en validation. Nous allons donc rester sur ce que nous avons de mieux.

# --------------------------------------------------------------------------------------------------------------